### DISCUSSION TITLE
Delivering Exe files - packaging signing and distribution

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#1021] [Q&A]
Hello,
For Windows, we are:
- compiling exe-bundles
- then we package the files up with Install Forge
- and the send the installer to the client.
We are wondering what service you guys use to sign the installers for Windows so that after downloading, Windows won't say it's an unknown publisher.
Curious in general what is your process for ditrubution of Windows apps.

Thanks!

**Comment by ajwfrost** [comment]
To avoid the Windows Defender messages, you need to have an "Extended Validation" certificate: normally it does some reputation-based checks, although you can also submit your app to Microsoft for them to add it to the approved list (but because we re-publish our .exe files frequently, we went with the EV signing approach...)

https://docs.microsoft.com/en-us/windows/security/threat-protection/microsoft-defender-smartscreen/microsoft-defender-smartscreen-overview
https://en.wikipedia.org/wiki/Extended_Validation_Certificate
https://answers.microsoft.com/en-us/protect/forum/all/defendersmartscreen-warning/bba4ad97-c145-4cad-9963-70782371cea9

We sign the .exe and .dll files and then package up the .msi and sign that too...

**Reply by raresn** [reply]
What tool do you use to sign the exe?
What about the tool you use to package up the msi? 

Thanks

**Reply by ajwfrost** [reply]
Microsoft's signtool:
`C:\Program Files (x86)\Windows Kits\10\App Certification Kit\signtool.exe`
using a certificate in PFX format.

To create the MSIs we use a bit of a mixture but mostly WiX - I'm not massively familiar with that part but I believe it comes as a plug-in to Visual Studio, or it uses Visual Studio, or similar..

**Reply by ajwfrost** [reply]
I know one of our ITSec guys was using a tool from Digicert, which is who issued our certificates, but I think that just wraps up the Microsoft tool..

**Reply by raresn** [reply]
This is very good info! Thank you very much, @ajwfrost !

**Reply by SUSDEV** [reply]
FWIW, since 2016, I have a development machine that is a VM on which I use unfuddle as an SVN repo, which (a) makes it easy to backup in case it gets screwed up (I've had several Windows Updates screw up the machine, and thankfully I can revert to prior machines then pull from the repo), and (b) allows me to upgrade my machines and continue developing.
I use FDT to compile my Flash/AIR work into a Windows exe, and I get a code signing cert from GoDaddy every few years. Then use InnoSetup to produce the installer, which has integration with kSign, [see here for instructions](https://jrsoftware.org/ishelp/index.php?topic=setup_signtool). I used Advanced Installer if I need to make an MSI.

Here is my notes file in all its glory, let me know if you need further explanations:

> I need to create 2 packages:
> 
> A. Windows 64-bit (raw application and exe installer)
> B. Windows 32-bit (raw application and exe installer)
> 
> Prep:
> Make sure version # is correct in the -app.xml files as well as in the installer (InnoSetup) and Advanced Installer scripts
> 
> Steps:
> 1. To make A & B, use the correct -app file, either -64 or -x86
> 2. Compile the app in FDT
> 3. Open the CMD prompt (run as admin) and use adt to create the captive runtime
> 
> cd "C:\Users\MyUserName\workspace\MyWorkSpace\bin"
> 
> "C:\Users\MyUserName\Documents\Flex-Air SDKs\AIR 33.1.1.554\bin\adt" -package -storetype pkcs12 -keystore myCertName.p12 -storepass "MyPassword" -tsa none -target bundle SimsUshare.exe SUSFreeWinMac-app.xml -C . SUSFreeWinMac.swf libs/ examples/ assets/ icons/ config/ piccache/ 
> 
> 4. Use the InnoSetup installer app to (a) sign the executable (in SimsUshare.exe/) and (b) package the exe and sign the installer.
> 5. Use the Advanced Installer to make the MSI files each for 32-bit and 64 bit.
> 6. I sign the application in the SimsUshare.exe/ folder because I will copy the folder out to use as the raw files when not installed by installer
> "C:\Program Files (x86)\kSign\signtool.exe" sign /f "C:\Users\MyUserName\workspace\MyWorkSpace\SUS Code Signing\Cert for Win.pfx" /t "http://timestamp.comodoca.com/?td=sha256" /p "MyPassword" "C:\Users\MyUserName\workspace\MyWorkSpace\bin\SimsUshare.exe\SimsUshare.exe"
> 
> 7. Lastly, I sign the installer (exe) and the MSI files manually ($f is each filename)
> "C:\Program Files (x86)\kSign\signtool.exe" sign /f "C:\Users\MyUserName\workspace\MyWorkSpace\SUS Code Signing\Cert for Win.pfx" /t "http://timestamp.comodoca.com/?td=sha256" /p "MyPassword" $f
> 
> Like:
> 
> "C:\Program Files (x86)\kSign\signtool.exe" sign /f "C:\Users\MyUserName\workspace\MyWorkSpace\SUS Code Signing\Cert for Win.pfx" /t "http://timestamp.comodoca.com/?td=sha256" /p "MyPassword" "C:\Users\MyUserName\workspace\MyWorkSpace\bin\Output\setup-simsushare.exe"
> 
> and
> 
> "C:\Program Files (x86)\kSign\signtool.exe" sign /f "C:\Users\MyUserName\workspace\MyWorkSpace\SUS Code Signing\Cert for Win.pfx" /t "http://timestamp.comodoca.com/?td=sha256" /p "MyPassword" "C:\Users\MyUserName\workspace\MyWorkSpace\bin\Output\SimsUshare.msi"

**Reply by SUSDEV** [reply]
While I'm at it, I wanted to add how I get the certificate into the right format, which I have to do every 3 years or so when I renew my certificate. Here are my notes from 2019, the last time I went through this.

It looks like I had received a certificate from GoDaddy but when I tried to use it, I could not get it into a form in which I could export a PFX file, which is what my computer needed. I even tried using the DigiCert tool which did not allow PFX export (:-().

So, I started from scratch.

Basically, in Windows 10, search for "user certificate" that will bring up where the certificates are stored (it looks like same as MMC in prior instructions).

I followed the steps ([from here](https://stackoverflow.com/questions/37443253/generating-certificate-signing-request-using-management-console))

> I ran into the same problem today and found the solution on MSDN. Try the following:
> 
> 1. Instead of selecting Active Directory Enrollment Policy select Proceed without enrollment policy.
> 2. Click Next.
> 3. Select (No template) CNG key from the Template list.
> 4. Select PKCS 10 as the Request format.
> 5. Click Next.
> 6. Click the Details arrow and then the Properties button.
> 7. Enter a name for your certificate in Friendly name box on the General tab.
> 8. Click the Subject tab.
> 9. Under Subject name, select Common name from the Type list. Enter a common name in the Value box.
> 10. Repeat step 9 for Organizational unit, Locality, State and Country.
> 11. Click the Extensions tab.
> 12. Under Key usage select Digital signature and click the Add button.
> 13. Under Extended key usage select Code signing and click the Add button.
> 14. Under Basic constraints click the Enable this extension checkbox.
> 15. Click the Private key tab.
> 16. Under Key options select 2048 as the Key size.
> 17. Click the Make private key exportable checkbox.
> 18. Under Select hash algorithm select sha256 from the Hash Algorithm list.
> 19. Click OK.
> 
> 
> So far it appears to be working. I was able to use my new code signing certificate to export a PFX file and successfully sign an executable.
> 
> Note: I chose sha256 instead of sha1 in step 18 because SHA-1 is deprecated.
>
> So far it appears to be working. I was able to use my new code signing certificate to export a PFX file and successfully sign an e executable.
> 
> Note: I chose sha256 instead of sha1 in step 18 because SHA-1 is deprecated.

With those steps, I generated a "SUSCert2019.req" (the CSR).

I then went to GoDaddy and had my certificate re-keyed, with the SUSCert2019.req value. I expect at that point to download some files from GoDaddy once issued then be able to bring those into my "Certificates (local computer)", and export the PFX.

It all seemed to work! The PFX (I needed to bring into MMC) and I copied it to P12 and that seems to work as is. For those wondering how to do this, [please see here](https://www.godaddy.com/help/windows-install-codedriver-signing-certificate-and-create-pfx-file-2698). Unfortunately, I see that GoDaddy no longer issues code signing certificates after June 1, 2021. I think mine expires in 2022, so I will have to find an alternative (any suggestions?).
